dashboard/app/models/puzzle_rating.rb

```diff
@@ -0,0 +1,43 @@
+# == Schema Information
+#
+# Table name: puzzle_ratings
+#
+#  id         :integer          not null, primary key
+#  user_id    :integer
+#  script_id  :integer
+#  level_id   :integer
+#  rating     :integer
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+# Indexes
+#
+#  index_puzzle_ratings_on_script_id_and_level_id              (script_id,level_id)
+#  index_puzzle_ratings_on_user_id                             (user_id)
+#  index_puzzle_ratings_on_user_id_and_script_id_and_level_id  (user_id,script_id,level_id) UNIQUE
+#
+
+class PuzzleRating < ActiveRecord::Base
+  belongs_to :user
+  belongs_to :script
+  belongs_to :level
+
+  validates :script, :presence => true
+  validates :level, :presence => true
+
+  validates_uniqueness_of :user_id, :scope => [:script_id, :level_id], :allow_nil => true
+
+  def PuzzleRating.enabled?
+    true
+  end
+
+  # If PuzzleRating is disabled, no one can rate
+  # If the user is not logged in, they can always rate
+  # If the user is logged in, they can only rate if
+  #   they haven't rated before
+  def PuzzleRating.can_rate?(script, level, user)
```

Comment:
I'm not talking about changing the data model, just the arguments of the
method.

On Tue, Oct 27, 2015 at 2:43 PM, ashercodeorg <notifications@github.com>
wrote:

> In dashboard/app/models/puzzle_rating.rb
> <https://github.com/code-dot-org/code-dot-org/pull/4753#discussion_r43191937>
> :
>
> > +  belongs_to :level
> > +
> > +  validates :script, :presence => true
> > +  validates :level, :presence => true
> > +
> > +  validates_uniqueness_of :user_id, :scope => [:script_id, :level_id], :allow_nil => true
> > +
> > +  def PuzzleRating.enabled?
> > +    true
> > +  end
> > +
> > +  # If PuzzleRating is disabled, no one can rate
> > +  # If the user is not logged in, they can always rate
> > +  # If the user is logged in, they can only rate if
> > +  #   they haven't rated before
> > +  def PuzzleRating.can_rate?(script, level, user)
>
> If ScriptLevel makes more sense, I'm happy to go that route. Though more
> difficult for me (necessitates downstream joins), it might be "better" per
> Laurel's comment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/4753/files#r43191937>.
>